baseapp/baseapp_test.go

```diff
@@ -2153,3 +2202,134 @@ func TestBaseApp_Init(t *testing.T) {
+	codec := codec.NewLegacyAmino()
+	registerTestCodec(codec)
+	numTransactions := len(txs)
+
+	app.BeginBlock(abci.RequestBeginBlock{Header: tmproto.Header{Height: blockHeight}})
+	for txNum := 0; txNum < numTransactions; txNum++ {
+		tx := txs[txNum]
+
+		txBytes, err := codec.Marshal(tx)
+		require.NoError(t, err)
+		resp := app.DeliverTx(abci.RequestDeliverTx{Tx: txBytes})
+		require.True(t, resp.IsOK(), "%v", resp.String())
+	}
+	app.EndBlock(abci.RequestEndBlock{Height: blockHeight})
+}
+
+func TestGenerateAndLoadFraudProof(t *testing.T) {
+	/*
+		Covers some parts of the cycle of a fraudproof by simulating the following steps:
+		1. Initialize a baseapp, B1, with some state, S0
+		2. Make some state transition to state S1 by doing some transactions
+		3. Export that state S1 into a fraudProof data structure (minimal snapshot)
+		4. Load a fresh baseapp, B2, with the contents of fraud proof data structure from (3) so it can begin from state S1.
+		5. Now, pick some set of transactions, txs1, to make some more state transitions.
+		6. Execute txs1 on both B1 and B2 so they go through the same state transitions
+		7. If the state of both B1 and B2 has to converge at a state S2, then we the test passes.
+
+		This test passing means cosmos-sdk can switch between a baseapp and fraudproof.
+
+		Tests to write in future:
+
+		1. Block with bad txs: Txs that exceed gas limits, validateBasic fails, unregistered messages (see TestRunInvalidTransaction)
+		2. Block with invalid appHash at the end
+		3. Corrupted Fraud Proof: bad SMT format, insufficient key-value pairs inside SMT needed to verify fraud
+		4. Bad block, fraud proof needed, fraud proof works, chain halts (happy case)
+
+		Notes: In the current implementation, all substores might not be SMTs, but we assume they are for the prototype here
+		Try to keep tx as generic as possible so you don't need to care about the messages inside a Tx
+
+	*/
+
+	storeTraceBuf := &bytes.Buffer{}
+	subStoreTraceBuf := &bytes.Buffer{}
+
+	routerOpt := func(bapp *BaseApp) {
+		bapp.Router().AddRoute(sdk.NewRoute(routeMsgKeyValue, func(ctx sdk.Context, msg sdk.Msg) (*sdk.Result, error) {
+			kv := msg.(*msgKeyValue)
+			bapp.cms.GetKVStore(capKey2).Set(kv.Key, kv.Value)
+			return &sdk.Result{}, nil
+		}))
+	}
+
+	// BaseApp, B1
+
+	appB1 := setupBaseApp(t,
+		AppOptionFunc(routerOpt),
+		SetSubstoreTracer(storeTraceBuf),
+		SetTracerFor(capKey2, subStoreTraceBuf),
+	)
+
+	appB1.InitChain(abci.RequestInitChain{})
+
+	// State here: S0
+
+	numTransactions := 50
+	executeBlockWithArbitraryTxs(t, appB1, numTransactions, 1)
+	appB1.Commit()
+
+	cms := appB1.cms.(*multi.Store)
+
+	// Exports all data inside current multistore into a fraudProof (S1) //
+
+	storeKeyToSubstoreTraceBuf := make(map[stypes.StoreKey]*bytes.Buffer)
+	storeKeyToSubstoreTraceBuf[capKey2] = subStoreTraceBuf
+
+	fraudProof := generateFraudProof(cms, storeKeyToSubstoreTraceBuf)
```

Comment:
Can you clarify a bit the order of operations to get to here? Does `appB1.Commit()` above commit state changes to disk? Does `generateFraudProof` get Merkle proofs against the post-execution root, or the pre-execution root?

Reply:
Calls to `Commit()` are irrelevant here, so I removed them for clarity. `app.cms` is able to access any state modified by block transactions. 

`generateFraudProof` gets Merkle proofs against the post-execution root. We'd like the pre-execution root which is where rolling back the app's state to the previous block is needed.